src/Entities/Entity.cpp

```diff
@@ -1559,68 +1536,87 @@ bool cEntity::DetectPortal()
 
 
 
-bool cEntity::DoMoveToWorld(cWorld * a_World, bool a_ShouldSendRespawn, Vector3d a_NewPosition)
+void cEntity::DoMoveToWorld(const sWorldChangeInfo & a_WorldChangeInfo)
 {
-	UNUSED(a_ShouldSendRespawn);
-	ASSERT(a_World != nullptr);
+	ASSERT(a_WorldChangeInfo.m_NewWorld != nullptr);
 
-	if (GetWorld() == a_World)
+	if (m_World != a_WorldChangeInfo.m_NewWorld)
 	{
-		// Don't move to same world
-		return false;
+		// Tell others we are gone
+		m_World->BroadcastDestroyEntity(*this);
 	}
 
-	//  Ask the plugins if the entity is allowed to changing the world
-	if (cRoot::Get()->GetPluginManager()->CallHookEntityChangingWorld(*this, *a_World))
+	#ifdef _DEBUG
+		// Take note of old chunk coords
+		auto OldChunkCoords = cChunkDef::BlockToChunk(GetPosition());
+	#endif
+
+	// Set position to the new position
+	SetPosition(a_WorldChangeInfo.m_NewPosition);
+
+	if (a_WorldChangeInfo.m_SetPortalCooldown)
 	{
-		// A Plugin doesn't allow the entity to changing the world
-		return false;
+		m_PortalCooldownData.m_TicksDelayed = 0;
+		m_PortalCooldownData.m_ShouldPreventTeleportation = true;
+	}
+
+	if (GetWorld() == a_WorldChangeInfo.m_NewWorld)
+	{
+		// Moving to same world, don't need to remove from world
+		return;
 	}
 
 	// Stop ticking, in preperation for detaching from this world.
 	SetIsTicking(false);
 
-	// Tell others we are gone
-	GetWorld()->BroadcastDestroyEntity(*this);
+	a_WorldChangeInfo.m_NewWorld->BroadcastSpawnEntity(*this);
```

Comment:
This line makes me uncomfortable.  It's locking the new world while the lock for the current is already held and thus a potential deadlock.

I think the cleanest way to fix this is to move the Broadcast call into `cWorld::Tick` when `m_EntitiesToAdd` is processed.  Anybody have thoughts on this?